{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "!pip install keras-tuner -q"
      ],
      "metadata": {
        "id": "VtmK-80n6Dch"
      },
      "execution_count": 46,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 47,
      "metadata": {
        "id": "V6wEDzTD3C-a"
      },
      "outputs": [],
      "source": [
        "import tensorflow as tf\n",
        "from tensorflow import keras\n",
        "from tensorflow.keras import layers\n",
        "from keras_tuner.tuners import Hyperband"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# create model\n",
        "def build_model(hp):\n",
        "  model=keras.Sequential()\n",
        "  # search num conv\n",
        "  num_conv_layers = hp.Int('num_conv_layers', min_value=3, max_value=5, step=1)\n",
        "  for i in range(num_conv_layers):\n",
        "    # search num filtwe\n",
        "    model.add(layers.Conv2D(\n",
        "        filters=hp.Int(f'conv_{i}_filters', min_value=32, max_value=256, step=32),\n",
        "        kernel_size=3,\n",
        "        activation='relu'\n",
        "    ))\n",
        "    if i%2==0:\n",
        "      model.add(layers.MaxPooling2D(pool_size=2))\n",
        "    model.add(layers.Dropout(rate=0.2))\n",
        "\n",
        "\n",
        "  model.add(layers.Flatten())\n",
        "  # search num dense\n",
        "  num_dense_layer=hp.Int('num_dense_layer',min_value=3, max_value=5, step=1)\n",
        "  for i in range(num_dense_layer):\n",
        "    # search num nerouns\n",
        "    model.add(layers.Dense(\n",
        "        units=hp.Int(f'dense_{i}_units', min_value=32, max_value=256, step=32),\n",
        "        activation='relu'\n",
        "    ))\n",
        "  model.add(layers.Dense(10, activation='softmax'))\n",
        "\n",
        "  # search learning rate\n",
        "  optimizer = keras.optimizers.Adam(\n",
        "      learning_rate=hp.Choice('learning_rate', values=[1e-1, 1e-2, 1e-3])\n",
        "  )\n",
        "  model.compile(optimizer=optimizer,\n",
        "              loss='categorical_crossentropy',\n",
        "              metrics=['accuracy'])\n",
        "  return model\n"
      ],
      "metadata": {
        "id": "Jcx0m7UZ6fI1"
      },
      "execution_count": 48,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# load dataset\n",
        "(x_train, y_train), (x_val, y_val) = keras.datasets.mnist.load_data()"
      ],
      "metadata": {
        "id": "SUfVLRJuJb-3"
      },
      "execution_count": 49,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "x_train = x_train.astype('float32') / 255.0\n",
        "x_val = x_val.astype('float32') / 255.0\n",
        "y_train = keras.utils.to_categorical(y_train)\n",
        "y_val = keras.utils.to_categorical(y_val)"
      ],
      "metadata": {
        "id": "1_cAbER3JkPx"
      },
      "execution_count": 50,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "x_train = np.expand_dims(x_train, -1)\n",
        "x_val = np.expand_dims(x_val, -1)"
      ],
      "metadata": {
        "id": "pjT0sz0B6X_q"
      },
      "execution_count": 51,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(x_train.shape)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SaJ-c1i16gBs",
        "outputId": "fee6a79e-d3ca-4e08-f77c-7db76e7805ce"
      },
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(60000, 28, 28, 1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "tuner = Hyperband(\n",
        "    build_model,\n",
        "    objective=\"val_accuracy\",\n",
        "    max_epochs=5,\n",
        "    executions_per_trial=2,\n",
        "    overwrite=True,\n",
        "    directory='tuner_dir',\n",
        "    project_name='mnist_model'\n",
        ")\n"
      ],
      "metadata": {
        "id": "zM8Yr7T7JpPd"
      },
      "execution_count": 55,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "tuner.search(x_train, y_train, epochs=10, validation_data=(x_val, y_val))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3vF0OiWgJwHc",
        "outputId": "42c46967-1f20-4853-f672-abb2655c2f88"
      },
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Trial 9 Complete [00h 02m 48s]\n",
            "val_accuracy: 0.5418500006198883\n",
            "\n",
            "Best val_accuracy So Far: 0.9918000102043152\n",
            "Total elapsed time: 00h 11m 30s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Get the top 2 models.\n",
        "models = tuner.get_best_models(num_models=2)\n",
        "best_model = models[0]\n",
        "# Build the model.\n",
        "# Needed for `Sequential` without specified `input_shape`.\n",
        "best_model.build(input_shape=(None, 28, 28,1))\n",
        "best_model.summary()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "F7Lyae8HJzcK",
        "outputId": "77362b4e-251c-4fe9-ef5f-baf42ba45c15"
      },
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Model: \"sequential\"\n",
            "_________________________________________________________________\n",
            " Layer (type)                Output Shape              Param #   \n",
            "=================================================================\n",
            " conv2d (Conv2D)             (None, 26, 26, 32)        320       \n",
            "                                                                 \n",
            " max_pooling2d (MaxPooling2  (None, 13, 13, 32)        0         \n",
            " D)                                                              \n",
            "                                                                 \n",
            " dropout (Dropout)           (None, 13, 13, 32)        0         \n",
            "                                                                 \n",
            " conv2d_1 (Conv2D)           (None, 11, 11, 96)        27744     \n",
            "                                                                 \n",
            " dropout_1 (Dropout)         (None, 11, 11, 96)        0         \n",
            "                                                                 \n",
            " conv2d_2 (Conv2D)           (None, 9, 9, 96)          83040     \n",
            "                                                                 \n",
            " max_pooling2d_1 (MaxPoolin  (None, 4, 4, 96)          0         \n",
            " g2D)                                                            \n",
            "                                                                 \n",
            " dropout_2 (Dropout)         (None, 4, 4, 96)          0         \n",
            "                                                                 \n",
            " flatten (Flatten)           (None, 1536)              0         \n",
            "                                                                 \n",
            " dense (Dense)               (None, 32)                49184     \n",
            "                                                                 \n",
            " dense_1 (Dense)             (None, 128)               4224      \n",
            "                                                                 \n",
            " dense_2 (Dense)             (None, 64)                8256      \n",
            "                                                                 \n",
            " dense_3 (Dense)             (None, 10)                650       \n",
            "                                                                 \n",
            "=================================================================\n",
            "Total params: 173418 (677.41 KB)\n",
            "Trainable params: 173418 (677.41 KB)\n",
            "Non-trainable params: 0 (0.00 Byte)\n",
            "_________________________________________________________________\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:tensorflow:Detecting that an object or model or tf.train.Checkpoint is being deleted with unrestored values. See the following logs for the specific values in question. To silence these warnings, use `status.expect_partial()`. See https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restorefor details about the status object returned by the restore function.\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-0.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-0.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-1.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-1.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-2.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-2.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-3.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-3.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-4.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-4.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-5.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-5.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-6.kernel\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).layer_with_weights-6.bias\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.1\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.2\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.3\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.4\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.5\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.6\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.7\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.8\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.9\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.10\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.11\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.12\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.13\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.14\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.15\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.16\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.17\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.18\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.19\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.20\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.21\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.22\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.23\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.24\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.25\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.26\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.27\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.28\n",
            "WARNING:tensorflow:Detecting that an object or model or tf.train.Checkpoint is being deleted with unrestored values. See the following logs for the specific values in question. To silence these warnings, use `status.expect_partial()`. See https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restorefor details about the status object returned by the restore function.\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.1\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.2\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.3\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.4\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.5\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.6\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.7\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.8\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.9\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.10\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.11\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.12\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.13\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.14\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.15\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.16\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.17\n",
            "WARNING:tensorflow:Value in checkpoint could not be found in the restored object: (root).optimizer._variables.18\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "tuner.results_summary()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "L88j-esoFVSL",
        "outputId": "ce9ed5b7-86e4-482f-f590-f451ccffee15"
      },
      "execution_count": 63,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Results summary\n",
            "Results in tuner_dir/mnist_model\n",
            "Showing 10 best trials\n",
            "Objective(name=\"val_accuracy\", direction=\"max\")\n",
            "\n",
            "Trial 0005 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 3\n",
            "conv_0_filters: 32\n",
            "conv_1_filters: 96\n",
            "conv_2_filters: 96\n",
            "num_dense_layer: 3\n",
            "dense_0_units: 32\n",
            "dense_1_units: 128\n",
            "dense_2_units: 64\n",
            "learning_rate: 0.001\n",
            "conv_3_filters: 128\n",
            "conv_4_filters: 96\n",
            "dense_3_units: 32\n",
            "dense_4_units: 160\n",
            "tuner/epochs: 5\n",
            "tuner/initial_epoch: 2\n",
            "tuner/bracket: 1\n",
            "tuner/round: 1\n",
            "tuner/trial_id: 0003\n",
            "Score: 0.9918000102043152\n",
            "\n",
            "Trial 0003 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 3\n",
            "conv_0_filters: 32\n",
            "conv_1_filters: 96\n",
            "conv_2_filters: 96\n",
            "num_dense_layer: 3\n",
            "dense_0_units: 32\n",
            "dense_1_units: 128\n",
            "dense_2_units: 64\n",
            "learning_rate: 0.001\n",
            "conv_3_filters: 128\n",
            "conv_4_filters: 96\n",
            "dense_3_units: 32\n",
            "dense_4_units: 160\n",
            "tuner/epochs: 2\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 1\n",
            "tuner/round: 0\n",
            "Score: 0.9878000020980835\n",
            "\n",
            "Trial 0002 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 4\n",
            "conv_0_filters: 64\n",
            "conv_1_filters: 32\n",
            "conv_2_filters: 96\n",
            "num_dense_layer: 3\n",
            "dense_0_units: 224\n",
            "dense_1_units: 64\n",
            "dense_2_units: 160\n",
            "learning_rate: 0.001\n",
            "conv_3_filters: 32\n",
            "conv_4_filters: 96\n",
            "dense_3_units: 32\n",
            "dense_4_units: 32\n",
            "tuner/epochs: 2\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 1\n",
            "tuner/round: 0\n",
            "Score: 0.9869000017642975\n",
            "\n",
            "Trial 0008 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 3\n",
            "conv_0_filters: 32\n",
            "conv_1_filters: 256\n",
            "conv_2_filters: 256\n",
            "num_dense_layer: 4\n",
            "dense_0_units: 96\n",
            "dense_1_units: 32\n",
            "dense_2_units: 160\n",
            "learning_rate: 0.01\n",
            "conv_3_filters: 64\n",
            "conv_4_filters: 128\n",
            "dense_3_units: 96\n",
            "dense_4_units: 160\n",
            "tuner/epochs: 5\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 0\n",
            "tuner/round: 0\n",
            "Score: 0.5418500006198883\n",
            "\n",
            "Trial 0000 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 4\n",
            "conv_0_filters: 256\n",
            "conv_1_filters: 224\n",
            "conv_2_filters: 128\n",
            "num_dense_layer: 3\n",
            "dense_0_units: 128\n",
            "dense_1_units: 96\n",
            "dense_2_units: 192\n",
            "learning_rate: 0.01\n",
            "tuner/epochs: 2\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 1\n",
            "tuner/round: 0\n",
            "conv_3_filters: 32\n",
            "Score: 0.11349999904632568\n",
            "\n",
            "Trial 0006 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 3\n",
            "conv_0_filters: 128\n",
            "conv_1_filters: 160\n",
            "conv_2_filters: 96\n",
            "num_dense_layer: 5\n",
            "dense_0_units: 224\n",
            "dense_1_units: 64\n",
            "dense_2_units: 192\n",
            "learning_rate: 0.1\n",
            "conv_3_filters: 64\n",
            "conv_4_filters: 128\n",
            "dense_3_units: 128\n",
            "dense_4_units: 32\n",
            "tuner/epochs: 5\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 0\n",
            "tuner/round: 0\n",
            "Score: 0.11349999904632568\n",
            "\n",
            "Trial 0004 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 3\n",
            "conv_0_filters: 64\n",
            "conv_1_filters: 256\n",
            "conv_2_filters: 256\n",
            "num_dense_layer: 5\n",
            "dense_0_units: 224\n",
            "dense_1_units: 192\n",
            "dense_2_units: 32\n",
            "learning_rate: 0.01\n",
            "conv_3_filters: 64\n",
            "conv_4_filters: 96\n",
            "dense_3_units: 256\n",
            "dense_4_units: 160\n",
            "tuner/epochs: 2\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 1\n",
            "tuner/round: 0\n",
            "Score: 0.10814999788999557\n",
            "\n",
            "Trial 0001 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 5\n",
            "conv_0_filters: 160\n",
            "conv_1_filters: 256\n",
            "conv_2_filters: 256\n",
            "num_dense_layer: 5\n",
            "dense_0_units: 224\n",
            "dense_1_units: 128\n",
            "dense_2_units: 160\n",
            "learning_rate: 0.01\n",
            "conv_3_filters: 160\n",
            "tuner/epochs: 2\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 1\n",
            "tuner/round: 0\n",
            "conv_4_filters: 32\n",
            "dense_3_units: 32\n",
            "dense_4_units: 32\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py\", line 273, in _try_run_and_update_trial\n",
            "    self._run_and_update_trial(trial, *fit_args, **fit_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py\", line 238, in _run_and_update_trial\n",
            "    results = self.run_trial(trial, *fit_args, **fit_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/tuners/hyperband.py\", line 427, in run_trial\n",
            "    return super().run_trial(trial, *fit_args, **fit_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py\", line 314, in run_trial\n",
            "    obj_value = self._build_and_fit_model(trial, *args, **copied_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py\", line 233, in _build_and_fit_model\n",
            "    results = self.hypermodel.fit(hp, model, *args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/hypermodel.py\", line 149, in fit\n",
            "    return model.fit(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
            "    raise e.with_traceback(filtered_tb) from None\n",
            "  File \"/tmp/__autograph_generated_filelbz166qb.py\", line 15, in tf__train_function\n",
            "    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)\n",
            "ValueError: in user code:\n",
            "\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1401, in train_function  *\n",
            "        return step_function(self, iterator)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1384, in step_function  **\n",
            "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1373, in run_step  **\n",
            "        outputs = model.train_step(data)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1150, in train_step\n",
            "        y_pred = self(x, training=True)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
            "        raise e.with_traceback(filtered_tb) from None\n",
            "\n",
            "    ValueError: Exception encountered when calling layer 'conv2d_4' (type Conv2D).\n",
            "    \n",
            "    Negative dimension size caused by subtracting 3 from 2 for '{{node sequential/conv2d_4/Conv2D}} = Conv2D[T=DT_FLOAT, data_format=\"NHWC\", dilations=[1, 1, 1, 1], explicit_paddings=[], padding=\"VALID\", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true](sequential/dropout_3/dropout/SelectV2, sequential/conv2d_4/Conv2D/ReadVariableOp)' with input shapes: [32,2,2,160], [3,3,160,32].\n",
            "    \n",
            "    Call arguments received by layer 'conv2d_4' (type Conv2D):\n",
            "      • inputs=tf.Tensor(shape=(32, 2, 2, 160), dtype=float32)\n",
            "\n",
            "\n",
            "\n",
            "Trial 0007 summary\n",
            "Hyperparameters:\n",
            "num_conv_layers: 5\n",
            "conv_0_filters: 224\n",
            "conv_1_filters: 128\n",
            "conv_2_filters: 192\n",
            "num_dense_layer: 5\n",
            "dense_0_units: 192\n",
            "dense_1_units: 64\n",
            "dense_2_units: 32\n",
            "learning_rate: 0.1\n",
            "conv_3_filters: 192\n",
            "conv_4_filters: 160\n",
            "dense_3_units: 224\n",
            "dense_4_units: 64\n",
            "tuner/epochs: 5\n",
            "tuner/initial_epoch: 0\n",
            "tuner/bracket: 0\n",
            "tuner/round: 0\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py\", line 273, in _try_run_and_update_trial\n",
            "    self._run_and_update_trial(trial, *fit_args, **fit_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py\", line 238, in _run_and_update_trial\n",
            "    results = self.run_trial(trial, *fit_args, **fit_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/tuners/hyperband.py\", line 427, in run_trial\n",
            "    return super().run_trial(trial, *fit_args, **fit_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py\", line 314, in run_trial\n",
            "    obj_value = self._build_and_fit_model(trial, *args, **copied_kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py\", line 233, in _build_and_fit_model\n",
            "    results = self.hypermodel.fit(hp, model, *args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/hypermodel.py\", line 149, in fit\n",
            "    return model.fit(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
            "    raise e.with_traceback(filtered_tb) from None\n",
            "  File \"/tmp/__autograph_generated_filelbz166qb.py\", line 15, in tf__train_function\n",
            "    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)\n",
            "ValueError: in user code:\n",
            "\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1401, in train_function  *\n",
            "        return step_function(self, iterator)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1384, in step_function  **\n",
            "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1373, in run_step  **\n",
            "        outputs = model.train_step(data)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1150, in train_step\n",
            "        y_pred = self(x, training=True)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
            "        raise e.with_traceback(filtered_tb) from None\n",
            "\n",
            "    ValueError: Exception encountered when calling layer 'conv2d_4' (type Conv2D).\n",
            "    \n",
            "    Negative dimension size caused by subtracting 3 from 2 for '{{node sequential/conv2d_4/Conv2D}} = Conv2D[T=DT_FLOAT, data_format=\"NHWC\", dilations=[1, 1, 1, 1], explicit_paddings=[], padding=\"VALID\", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true](sequential/dropout_3/dropout/SelectV2, sequential/conv2d_4/Conv2D/ReadVariableOp)' with input shapes: [32,2,2,192], [3,3,192,160].\n",
            "    \n",
            "    Call arguments received by layer 'conv2d_4' (type Conv2D):\n",
            "      • inputs=tf.Tensor(shape=(32, 2, 2, 192), dtype=float32)\n",
            "\n",
            "\n"
          ]
        }
      ]
    }
  ]
}